third_party/terraform/resources/resource_bigquery_table.go

```diff
@@ -132,6 +128,164 @@ func bigQueryTableSchemaDiffSuppress(_, old, new string, _ *schema.ResourceData)
+	switch old.(type) {
+	case []interface{}:
+		arrayOld := old.([]interface{})
+		arrayNew, ok := new.([]interface{})
+		if !ok {
+			// if not both arrays not changeable
+			return false, nil
+		} else if len(arrayOld) > len(arrayNew) {
+			// if not growing not changeable
+			return false, nil
+		}
+		for i := range arrayOld {
+			isChangable, err := resourceBigQueryTableSchemaIsChangable(arrayOld[i], arrayNew[i])
+			if err != nil || !isChangable {
+				return false, err
+			}
+		}
+		return true, nil
+	case map[string]interface{}:
+		objectOld := old.(map[string]interface{})
+		objectNew, ok := new.(map[string]interface{})
+		if !ok {
+			// if both aren't objects
+			return false, nil
+		}
+
+		var unionOfKeys map[string]bool = make(map[string]bool)
+		for key := range objectOld {
+			unionOfKeys[key] = true
+		}
+		for key := range objectNew {
+			unionOfKeys[key] = true
+		}
+
+		for key := range unionOfKeys {
+			valOld := objectOld[key]
+			valNew := objectNew[key]
+			switch key {
+			case "name":
+				if valOld != valNew {
+					return false, nil
+				}
+			case "type":
+				if !bigQueryTableTypeEq(valOld, valNew) {
+					return false, nil
+				}
+			case "mode":
+				if bigQueryTableModeIsForceNew(valOld, valNew) {
+					return false, nil
+				}
+			case "fields":
+				return resourceBigQueryTableSchemaIsChangable(valOld, valNew)
+
+				// other parameters: description, policyTags and
+				// policyTags.names[] are changeable
+			}
+		}
+		return true, nil
+	case string, float64, bool, nil:
+		// realistically this shouldn't hit
+		log.Printf("[DEBUG] comparison of generics hit... not expected")
+		return old == new, nil
+	default:
+		log.Printf("[DEBUG] tried to iterate through json but encountered a non native type to json deserialization... please ensure you are passing a json object from json.Unmarshall")
+		return false, errors.New("unable to compare values")
+	}
+}
+
+func resourceBigQueryTableSchemaCustomizeDiffFunc(d TerraformResourceDiff) error {
+	if _, hasSchema := d.GetOk("schema"); hasSchema {
+		if _, hasfield := d.GetOk("field.#"); hasfield {
+			if err := d.ForceNew("schema"); err != nil {
+				return err
+			}
+			if err := d.ForceNew("field.#"); err != nil {
+				return err
+			}
+		} else {
+			oldSchema, newSchema := d.GetChange("schema")
+			oldSchemaText := oldSchema.(string)
+			newSchemaText := newSchema.(string)
+			if oldSchemaText == "null" {
+				oldSchemaText = "[]"
+			}
+			var old, new interface{}
+			if err := json.Unmarshal([]byte(oldSchemaText), &old); err != nil {
+				log.Printf("[DEBUG] unable to unmarshal json customized diff - %v", err)
+			}
+			if err := json.Unmarshal([]byte(newSchemaText), &new); err != nil {
+				log.Printf("[DEBUG] unable to unmarshal json customized diff - %v", err)
+			}
+			isChangable, err := resourceBigQueryTableSchemaIsChangable(old, new)
+			if err != nil {
+				return err
+			} else if !isChangable {
```

Comment:
nit: you can remove the `else` because you returned early in the prior `if`. See https://golang.org/doc/effective_go.html#if as well as a great example at https://danp.net/2015/11/02/reducing-go-nesting.html.

Reply:
thanks for the reads :) updated